### PR TITLE
Add base for build job using older GCC version

### LINKED ISSUE #1058: Use old GCC for CI
We should think about having older versions of gcc compile us in CI or buildbot at some point!

_Originally posted by @CohenArthur in https://github.com/Rust-GCC/gccrs/pull/1057#pullrequestreview-919252016_

### PR DESCRIPTION
**Pull request by CohenArthur**
Fixes #1058

### CODE REVIEW

**Comment by CohenArthur** [comment]
So I guess this is proving to be useful! 

```
In file included from ../../gcc/rust/ast/rust-ast-full.h:22:0,
                 from ../../gcc/rust/parse/rust-parse.h:21,
                 from ../../gcc/rust/rust-session-manager.cc:23:
../../gcc/rust/ast/rust-ast.h: In member function ‘Rust::AST::LifetimeParam Rust::Parser<ManagedTokenSource>::parse_lifetime_param() [with ManagedTokenSource = Rust::Lexer]’:
../../gcc/rust/ast/rust-ast.h:1173:7: warning: ‘lifetime.Rust::AST::Lifetime::node_id’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 class Lifetime : public TypeParamBound
       ^
In file included from ../../gcc/rust/parse/rust-parse.h:722:0,
                 from ../../gcc/rust/rust-session-manager.cc:23:
../../gcc/rust/parse/rust-parse-impl.h:3331:17: note: ‘lifetime.Rust::AST::Lifetime::node_id’ was declared here
   AST::Lifetime lifetime (AST::Lifetime::NAMED, lifetime_tok->get_str (),
                 ^
In file included from ../../gcc/rust/ast/rust-ast-full.h:22:0,
                 from ../../gcc/rust/ast/rust-ast-full-test.cc:23:
../../gcc/rust/ast/rust-ast.h: In member function ‘Rust::AST::LifetimeParam Rust::Parser<ManagedTokenSource>::parse_lifetime_param() [with ManagedTokenSource = Rust::Lexer]’:
../../gcc/rust/ast/rust-ast.h:1173:7: warning: ‘lifetime.Rust::AST::Lifetime::node_id’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 class Lifetime : public TypeParamBound
       ^
In file included from ../../gcc/rust/parse/rust-parse.h:722:0,
                 from ../../gcc/rust/ast/rust-ast-full-test.cc:28:
../../gcc/rust/parse/rust-parse-impl.h:3331:17: note: ‘lifetime.Rust::AST::Lifetime::node_id’ was declared here
   AST::Lifetime lifetime (AST::Lifetime::NAMED, lifetime_tok->get_str (),
   ```
and
```
In file included from ../../gcc/rust/typecheck/rust-hir-type-check-type.h:25:0,
                 from ../../gcc/rust/typecheck/rust-hir-trait-resolve.h:25,
                 from ../../gcc/rust/backend/rust-compile.cc:19:
../../gcc/rust/typecheck/rust-hir-path-probe.h: In member function ‘void Rust::Resolver::PathProbeType::process_associated_trait_for_candidates(const Rust::Resolver::TraitReference*, Rust::HIR::ImplBlock*, bool)’:
../../gcc/rust/typecheck/rust-hir-path-probe.h:370:28: error: no matching function for call to ‘Rust::Resolver::PathProbeCandidate::PathProbeCandidate(<brace-enclosed initializer list>)’
      {trait_item_candidate}};
                            ^
../../gcc/rust/typecheck/rust-hir-path-probe.h:370:28: note: candidates are:
../../gcc/rust/typecheck/rust-hir-path-probe.h:91:3: note: Rust::Resolver::PathProbeCandidate::PathProbeCandidate(Rust::Resolver::PathProbeCandidate::CandidateType, Rust::TyTy::BaseType*, Location, Rust::Resolver::PathProbeCandidate::TraitItemCandidate)
   PathProbeCandidate (CandidateType type, TyTy::BaseType *ty, Location locus,
   ^
../../gcc/rust/typecheck/rust-hir-path-probe.h:91:3: note:   no known conversion for argument 4 from ‘<brace-enclosed initializer list>’ to ‘Rust::Resolver::PathProbeCandidate::TraitItemCandidate’
../../gcc/rust/typecheck/rust-hir-path-probe.h:86:3: note: Rust::Resolver::PathProbeCandidate::PathProbeCandidate(Rust::Resolver::PathProbeCandidate::CandidateType, Rust::TyTy::BaseType*, Location, Rust::Resolver::PathProbeCandidate::ImplItemCandidate)
   PathProbeCandidate (CandidateType type, TyTy::BaseType *ty, Location locus,
   ^
../../gcc/rust/typecheck/rust-hir-path-probe.h:86:3: note:   no known conversion for argument 4 from ‘<brace-enclosed initializer list>’ to ‘Rust::Resolver::PathProbeCandidate::ImplItemCandidate’
../../gcc/rust/typecheck/rust-hir-path-probe.h:81:3: note: Rust::Resolver::PathProbeCandidate::PathProbeCandidate(Rust::Resolver::PathProbeCandidate::CandidateType, Rust::TyTy::BaseType*, Location, Rust::Resolver::PathProbeCandidate::EnumItemCandidate)
   PathProbeCandidate (CandidateType type, TyTy::BaseType *ty, Location locus,
   ^
../../gcc/rust/typecheck/rust-hir-path-probe.h:81:3: note:   no known conversion for argument 4 from ‘<brace-enclosed initializer list>’ to ‘Rust::Resolver::PathProbeCandidate::EnumItemCandidate’
../../gcc/rust/typecheck/rust-hir-path-probe.h:31:8: note: constexpr Rust::Resolver::PathProbeCandidate::PathProbeCandidate(const Rust::Resolver::PathProbeCandidate&)
 struct PathProbeCandidate
        ^
../../gcc/rust/typecheck/rust-hir-path-probe.h:31:8: note:   candidate expects 1 argument, 4 provided
../../gcc/rust/typecheck/rust-hir-path-probe.h:31:8: note: constexpr Rust::Resolver::PathProbeCandidate::PathProbeCandidate(Rust::Resolver::PathProbeCandidate&&)
../../gcc/rust/typecheck/rust-hir-path-probe.h:31:8: note:   candidate expects 1 argument, 4 provided
../../gcc/rust/typecheck/rust-hir-path-probe.h: In member function ‘void Rust::Resolver::PathProbeType::process_predicate_for_candidates(const Rust::TyTy::TypeBoundPredicate&, bool)’:
../../gcc/rust/typecheck/rust-hir-path-probe.h:415:28: error: no matching function for call to ‘Rust::Resolver::PathProbeCandidate::PathProbeCandidate(<brace-enclosed initializer list>)’
      {trait_item_candidate}};
                            ^
../../gcc/rust/typecheck/rust-hir-path-probe.h:415:28: note: candidates are:
../../gcc/rust/typecheck/rust-hir-path-probe.h:91:3: note: Rust::Resolver::PathProbeCandidate::PathProbeCandidate(Rust::Resolver::PathProbeCandidate::CandidateType, Rust::TyTy::BaseType*, Location, Rust::Resolver::PathProbeCandidate::TraitItemCandidate)
   PathProbeCandidate (CandidateType type, TyTy::BaseType *ty, Location locus,
   ^
../../gcc/rust/typecheck/rust-hir-path-probe.h:91:3: note:   no known conversion for argument 4 from ‘<brace-enclosed initializer list>’ to ‘Rust::Resolver::PathProbeCandidate::TraitItemCandidate’
../../gcc/rust/typecheck/rust-hir-path-probe.h:86:3: note: Rust::Resolver::PathProbeCandidate::PathProbeCandidate(Rust::Resolver::PathProbeCandidate::CandidateType, Rust::TyTy::BaseType*, Location, Rust::Resolver::PathProbeCandidate::ImplItemCandidate)
   PathProbeCandidate (CandidateType type, TyTy::BaseType *ty, Location locus,
   ^
../../gcc/rust/typecheck/rust-hir-path-probe.h:86:3: note:   no known conversion for argument 4 from ‘<brace-enclosed initializer list>’ to ‘Rust::Resolver::PathProbeCandidate::ImplItemCandidate’
../../gcc/rust/typecheck/rust-hir-path-probe.h:81:3: note: Rust::Resolver::PathProbeCandidate::PathProbeCandidate(Rust::Resolver::PathProbeCandidate::CandidateType, Rust::TyTy::BaseType*, Location, Rust::Resolver::PathProbeCandidate::EnumItemCandidate)
   PathProbeCandidate (CandidateType type, TyTy::BaseType *ty, Location locus,
   ^
../../gcc/rust/typecheck/rust-hir-path-probe.h:81:3: note:   no known conversion for argument 4 from ‘<brace-enclosed initializer list>’ to ‘Rust::Resolver::PathProbeCandidate::EnumItemCandidate’
../../gcc/rust/typecheck/rust-hir-path-probe.h:31:8: note: constexpr Rust::Resolver::PathProbeCandidate::PathProbeCandidate(const Rust::Resolver::PathProbeCandidate&)
 struct PathProbeCandidate
        ^
../../gcc/rust/typecheck/rust-hir-path-probe.h:31:8: note:   candidate expects 1 argument, 4 provided
../../gcc/rust/typecheck/rust-hir-path-probe.h:31:8: note: constexpr Rust::Resolver::PathProbeCandidate::PathProbeCandidate(Rust::Resolver::PathProbeCandidate&&)
../../gcc/rust/typecheck/rust-hir-path-probe.h:31:8: note:   candidate expects 1 argument, 4 provided
```

Which seem small enough that we should fix them in this PR. I'll rebase for now so we have just one commit for the addition to the CI

**Review by tschwinge — APPROVED** [review]
See discussion in #1058.

I'm thinking about doing some changes here, but shall do that as individual separate Pull Requests on top of this.

---

I'm confirming that "hir-path-probe: Fix miscompilation on gcc-4.8" commit does fix that issue with GCC 4.8 -- that I had also run into, but not yet analyzed.

File: .github/workflows/ccpp.yml [hunk 1]
```diff
@@ -84,3 +84,45 @@ jobs:
...
+               CC='gcc-4.8' \
+               CXX='g++-4.8' \
+               --enable-languages=rust \
+               --disable-bootstrap \
+               --enable-multilib
+
+    - name: Build
+      shell: bash
+      run: |
+           make -C gccrs-build -j $(nproc)
```
Comment (philberty) [inline .github/workflows/ccpp.yml:128]: lets run the testsuite as well, but we don't need to exit on any unexpected failures

**Review by philberty — COMMENTED** [review]
does this job get run in parallel with the existing job?

**Comment by philberty** [comment]
I guess this job fails the build on any build failure with gcc4.8?

File: .github/workflows/ccpp.yml [hunk 2]
```diff
@@ -84,3 +84,45 @@ jobs:
             else \
               exit 0; \
             fi
+
+  build-and-check-older-gcc:
```
Comment (philberty) [inline .github/workflows/ccpp.yml:88]: I think this should be build-and-check-gcc4.8

**Comment by CohenArthur** [comment]
> does this job get run in parallel with the existing job?

Yes! It runs at the same time and finishes a little earlier too I believe

File: .github/workflows/ccpp.yml (hunk 1 above)
Comment (tschwinge) [reply .github/workflows/ccpp.yml:128]: Why not check the results?  This is quick, and should never ever differ from the main *build-and-check* job -- and if it ever does, there's a big problem that we'd like to learn about.

Comment (philberty) [reply .github/workflows/ccpp.yml:128]: Actually yeah good point if we ignore any failures we won't notice anything good point. Yeah lets copy this over from the normal job.

Comment (CohenArthur) [reply .github/workflows/ccpp.yml:128]: Done in [74078a6](https://github.com/Rust-GCC/gccrs/pull/1059/commits/74078a6242f73bb3c11d2985f71c30e5f2d1eb72)

File: .github/workflows/ccpp.yml (hunk 2 above)
Comment (CohenArthur) [reply .github/workflows/ccpp.yml]: Fixed in [74078a6](https://github.com/Rust-GCC/gccrs/pull/1059/commits/74078a6242f73bb3c11d2985f71c30e5f2d1eb72)

**Review by philberty — APPROVED** [review]
nice one

**Pull request merged**